core/network/src/on_demand.rs

```diff
@@ -316,6 +324,45 @@ impl<B> OnDemandService<B> for OnDemand<B> where
 			data => Accept::Unexpected(data),
 		})
 	}
+
+	fn on_remote_body_response(&self, peer: PeerId, response: message::BlockResponse<B>) {
+		self.accept_response("body", peer, response.id, |request| match request.data {
+			RequestData::RemoteBody(request, sender) => {
+				let num_bodies = response.blocks.iter().count();
+
+				// Number of bodies are hardcoded to 1 for valid `RemoteBodyResponses`
+				if num_bodies != 1 {
+					return Accept::CheckFailed("RemoteBodyResponse: invalid number of blocks".into(),
+												RequestData::RemoteBody(request, sender))
+				}
+
+				let response = response
+					.blocks
+					.into_iter()
+					.take(1)
+					.map(|b| match (b.body, b.header) {
+						(Some(b), Some(h)) => Some((b, h)),
+						_ => None
+					}).nth(0);
+
+				// Body and Header should included in valid `RemoteBodyResponses`
+				let (body, header) = match response {
+					Some(Some((b, h))) => (b, h),
+					_ => return Accept::CheckFailed("RemoteBodyResponse: is missing body or header".into(),
+													RequestData::RemoteBody(request, sender)),
+				};
+
+				if request.header.extrinsics_root() == header.extrinsics_root() {
```

Comment:
Think there's a problem with this check - it doesn't actually guarantee that the returned body matches the header we have requested. It only checks that extrinsic root is matched => this way full node could return header the light node expects + any body. Or am I missing something?

I'd personally in favor of fetching only `BODY` + computing extrinsic root using `HashFor::<Block>::ordered_trie_root(body.iter().map(Encode::encode))` + checking it against local `header.extrinsic_root()`. But this will only work if runtime actually puts that value to the `Header::extrisic_root`. Right now iiuc there are no assumptions of that in the native code, except for single `debug_assert_eq` in `BlockBuilder`. So probably someone else have an idea if we could rely on that fact? BTW - this way we could remove extra network messages && treat `BlockResponse` with only body filled as a response for `on_demand` - iirc Andre did the same with justification requests.

Reply:
Sounds much better to me, but I'm not familiar with the runtime

Reply:
@svyatonik based on Gav's response, shall we put this on ice or just compare `request.header.hash() == response.header.hash()` for now?

Reply:
IMO let's do as I've proposed ^^^ && file issue that current native (light client) code doesn't support situation when runtime returns anything else (i.e. not `ordered_trie_root(extrinsics)`) from `extrinsic_root()`. I see no other way how we could solve this issue. Committing vulnerable code IMO isn't an option.

Reply:
We can solve this at compile-time by requiring `extrinsic_root` to always return `ordered_trie_root(extrinsics)`, unless a Cargo feature is specified that disables light client support.